Synchronizing changes from Boeing to open source

BA-MBSE opened the pull request:
Update ElementsResponse to include commit id when committing changes
Branch id input sanitization
Add option to disable TWC auth delegation & host parsing cleanup
Added TokenService to allow use of tokens in other contexts
Allow password setting for admins (user conversion from external to internal)
Handle setting of null passwords (user conversion from internal to external)

BA-MBSE committed:
Update ElementsResponse to include commit id when committing changes
Branch id input sanitization
Add option to disable TWC auth delegation & host parsing cleanup
Added TokenService to allow use of tokens in other contexts
Allow password setting for admins (user conversion from external to internal)
Handle setting of null passwords (user conversion from internal to external)

HuiJun commented on rdb/src/main/java/org/openmbee/mms/rdb/repositories/BaseDAOImpl.java:
```diff
@@ -36,6 +38,10 @@ public JdbcTemplate getConn() {
     public String getSuffix() {
         String refId = ContextHolder.getContext().getBranchId();
-        return refId.equals(ContextObject.MASTER_BRANCH) ? "" : refId.toLowerCase();
+        if(Pattern.matches("^[a-zA-Z0-9-_]*$", refId)) { 
```
getSuffix() is used in a number of places multiple times in our DAOs. Recompiling the pattern every time could potentially have an effect on performance, albeit, probably not significantly.

dlamoris replied:
there's a BRANCH_ID_VALID_PATTERN in the BranchesController - may be better if that gets moved to core Formats?

HuiJun replied:
i think so

BA-MBSE replied:
updated to use a common Pattern

HuiJun commented on twc/src/main/java/org/openmbee/mms/twc/config/TwcConfig.java:
```diff
@@ -39,11 +51,21 @@ public TwcAuthenticationProvider getAuthNProvider(String associatedTWC) {
     }
+
+    private static String stripHost(String url) {
+        Pattern pattern = Pattern.compile("(https?://)?([\\w-\\.]*)(:\\d+)?");
```
Should check this and possibly move it to a static final.

dlamoris commented on core/src/main/java/org/openmbee/mms/core/objects/ElementsCommitResponse.java:
```diff
@@ -0,0 +1,13 @@
+package org.openmbee.mms.core.objects;
+
+public class ElementsCommitResponse extends ElementsResponse {
+    private String commitId;
```
add @Schema(nullable = true) for openapi

BA-MBSE replied:
added

BA-MBSE committed:
Review feedback.
Moved BRANCH_ID_VALID_PATTERN to core Constants & used both places
Added @Schema annotation to ElementsCommitResponse

HuiJun commented on rdb/src/main/java/org/openmbee/mms/rdb/repositories/BaseDAOImpl.java:
```diff
@@ -38,7 +40,7 @@ public JdbcTemplate getConn() {
     public String getSuffix() {
         String refId = ContextHolder.getContext().getBranchId();
-        if(Pattern.matches("^[a-zA-Z0-9-_]*$", refId)) { 
+        if(Pattern.matches(BRANCH_ID_VALID_PATTERN, refId)) {
```
From what I understand Pattern.matches(String, String) is a convenience method that does Pattern.matches(Pattern.compile(String), String) for regex patterns that are used only once. The expensive bit is Pattern.compile, which should instead be defined as static final so that it is only compiled once at initialization.

BA-MBSE replied:
Changed to a static Pattern instead

BA-MBSE committed:
Changed BRANCH_ID_VALID_PATTERN to a Pattern instead of a String
